tests/test_utils.py

```diff
@@ -102,3 +104,150 @@ class MyObj:
+    assert VersionNumber("dev") > VersionNumber("999999")
+
+    assert VersionNumber(1) <= VersionNumber("dev")
+    assert VersionNumber(999999) <= VersionNumber("dev")
+    assert VersionNumber("dev") >= VersionNumber("999999")
+
+    assert VersionNumber(1) != VersionNumber("dev")
+    assert not (VersionNumber(1) == VersionNumber("dev"))
+
+    with pytest.raises(ValueError):
+        assert VersionNumber("dev") > VersionNumber("-1")
+
+    with pytest.raises(ValueError):
+        assert VersionNumber("deva")
+
+
+def test_semantic_version_definition():
+    assert SemanticVersion((1, 2, 3))
+    assert SemanticVersion(("1", "2", "dev"))
+    assert SemanticVersion((1, "2", "dev"))
+
+    assert SemanticVersion(major=1, minor=2, patch=3)
+    assert SemanticVersion(major="1", minor="2", patch="3")
+    assert SemanticVersion(major=1, minor="2", patch="dev")
+    assert SemanticVersion(major=1, minor="2", patch="dev01")
+
+    assert SemanticVersion("1.1.dev")
+    assert SemanticVersion("1.1.dev1")
+
+    with pytest.raises(ValueError):
+        SemanticVersion((1, 2))
+
+    with pytest.raises(ValueError):
+        SemanticVersion(major=1, minor=2)
+
+    with pytest.raises(ValueError):
+        SemanticVersion(major=1, minor=2, patch="dev.a")
+
+    with pytest.raises(ValueError):
+        SemanticVersion(major=1, minor="dev", patch="dev1")
+
+    with pytest.raises(ValueError):
+        SemanticVersion("1.1.deva")
+
+    with pytest.raises(ValueError):
+        SemanticVersion("1.1")
+
+
+def test_semantic_version_methods():
+    ver = SemanticVersion("1.1.dev1")
+    assert ver.major == 1
+    assert ver.minor == 1
+    assert ver.patch == "dev1"
+
+    assert ver.as_string() == "1.1.dev1"
+    assert ver.as_tuple() == (1, 1, "dev1")
+    assert ver.as_dict() == {"major": 1, "minor": 1, "patch": "dev1"}
+    assert ver.as_list() == [1, 1, "dev1"]
+
+
+def test_semantic_version_comparison():
+    assert SemanticVersion("1.1.1") < SemanticVersion((1, 1, 2))
+    assert SemanticVersion("1.1.1") > SemanticVersion((1, 1, 0))
+
+    assert SemanticVersion("1.1.1") <= SemanticVersion((1, 1, 2))
+    assert SemanticVersion("1.1.1") >= SemanticVersion((1, 1, 0))
+
+    assert SemanticVersion("1.1.1") <= SemanticVersion((1, 1, 1))
+    assert SemanticVersion("1.1.1") >= SemanticVersion((1, 1, 1))
+
+    assert SemanticVersion("1.1.2") == SemanticVersion((1, 1, 2))
+    assert SemanticVersion("1.1.2") != SemanticVersion((1, 1, 3))
+
+    # checking major minors
+    assert SemanticVersion("1.1.1") < SemanticVersion((1, 2, 1))
+    assert SemanticVersion("1.1.1") > SemanticVersion((1, 0, 1))
+
+    assert SemanticVersion("1.1.1") < SemanticVersion((2, 1, 1))
+    assert SemanticVersion("1.1.1") > SemanticVersion((0, 1, 1))
+
+
+def test_semantic_version_comparison_dev():
+    assert SemanticVersion("1.1.1") < SemanticVersion((1, 1, "dev"))
+    assert SemanticVersion("1.1.1") < SemanticVersion((1, 1, "dev1"))
```

Comment:
I know it might sound silly, but order matters as well when overriding the comparison operators. Checking it both ways should confirm that it is working fine.

```suggestion
    assert SemanticVersion("1.1.1") < SemanticVersion((1, 1, "dev"))
    assert SemanticVersion("1.1.1") < SemanticVersion((1, 1, "dev1"))
    assert SemanticVersion((1, 1, "dev")) >  SemanticVersion("1.1.1")
    assert SemanticVersion((1, 1, "dev1")) > SemanticVersion("1.1.1")
```

Reply:
At some point I thought about this... but then it slip out my mind....